qutip/solver/parallel.py

```diff
@@ -194,20 +191,33 @@ def parallel_map(task, values, task_args=None, task_kwargs=None,
     finished = []
     if reduce_func is not None:
         results = None
-        result_func = lambda i, value: reduce_func(value)
+        def result_func(_, value):
+            return reduce_func(value)
     else:
         results = [None] * len(values)
-        result_func = lambda i, value: results.__setitem__(i, value)
+        result_func = results.__setitem__
 
     def _done_callback(future):
         if not future.cancelled():
             try:
                 result = future.result()
+                remaining_ntraj = result_func(future._i, result)
+                if remaining_ntraj is not None and remaining_ntraj <= 0:
+                    finished.append(True)
             except Exception as e:
                 errors[future._i] = e
-        remaining_ntraj = result_func(future._i, result)
-        if remaining_ntraj is not None and remaining_ntraj <= 0:
-            finished.append(True)
+            except KeyboardInterrupt:
```

Comment:
If raising `KeyboardInterrupt` in the main thread always causes a deadlock, then catching it here would seem to leave a small window somewhere outside of this `try ... except` and outside of this function where it might not be caught and the deadlock would happen anyway?

This definitely narrows the window where deadlock can occur, but could we do better or even eliminate the deadlock entirely?

Thank you for the great write up in the comment. I wouldn't have been able to ask this question without it.

Reply:
I don't really understand why that deadlock happens, but I do believe it should be gone now. What happens is this:
1. User presses CTRL+C -> KeyboardInterrupt is immediately raised in the main process and all worker processes
2. The worker processes return, and the corresponding `Future` objects record that the tasks didn't yield a result, but raised an exception.
3. In the main process, on the way out after the KeyboardInterrupt is raised, we hit the `finally` clause that shuts down the executor. The executor waits for all worker processes to finish and still calls the "done callback" for all the tasks that were currently running.
4. When `future.result()` is invoked in the callback, it raises the KeyboardInterrupt that was recorded in step 2 in the main thread. This is what causes that weird "deadlock".

The point is that the only problematic KeyboardInterrupt is the one that is raised from `future.result()`.

Reply:
Ah, then there is a small time window during which the `except KeyboardInterrupt` will catch the interrupt from the main thread instead of the raised by `future.result()`. To be completely safe we should not use only `.result()` but also check `future.exception()` (and pass if it is a KeyboardInterrupt).

Reply:
You are completely right, thank you! I have updated this part so that it checks `future.exception()`.

Reply:
Thanks!